jolt/src/parse.rs

```diff
@@ -82,13 +84,16 @@ fn parse_instructions(elf: &ElfBytes<LittleEndian>, data: &[u8]) -> Result<Vec<I
             });
 
             // UNIMP instruction is OK as long as it's not executed.
-            if inst.inst == RV32::ECALL || inst.inst == RV32::EBREAK {
-                tracing::debug!(
-                    target: LOG_TARGET,
-                    ?addr,
-                    "Unsupported instruction",
-                );
-                return Err(Error::Unsupported(inst.inst));
+            match inst.inst {
+                RV32::ECALL { .. } | RV32::EBREAK { .. } => {
+                    tracing::debug!(
+                        target: LOG_TARGET,
+                        ?addr,
+                        "Unsupported instruction",
+                    );
+                    return Err(Error::Unsupported(inst.inst));
```

Comment:
Based on a discussion with @sjudson, I anticipated the Jolt prover wouldn't fully work yet, at least not with `ECALL`s, but I also wanted to test and document the latest behavior.

I tried out the Jolt prover using the steps described at https://github.com/nexus-xyz/nexus-zkvm/pull/159#issuecomment-2120903704, except with `--impl jolt` since it's no longer the default, and on the HEAD of this branch instead, currently adc3886.

The default "Hello, World" program fails with the following error:

```shell
$ cargo nexus prove --impl jolt
   Compiling proc-macro2 v1.0.84
   Compiling unicode-ident v1.0.12
   Compiling syn v1.0.109
   Compiling nexus-rt v1.0.0 (.../nexus-zkvm/only-riscv/runtime)
   Compiling quote v1.0.36
   Compiling nexus-rt-macros v0.1.0 (.../nexus-zkvm/only-riscv/runtime/macros)
   Compiling pr-164 v0.1.0 (/private/tmp/pr-164)
    Finished release-unoptimized [unoptimized] target(s) in 3.03s
Error: Instruction isn't supported: ecall x10
```

Based on the error message, it must be coming from the highlighted line. Again, this isn't surprising and on its own doesn't necessarily warrant blocking the PR since Jolt isn't the default prover.

Reply:
When I removed any source of `ECALL`s in the program to prove/verify, the `jolt` impl worked as expected. I've not audited to see what, if anything, the optimizer elided as dead code, but it's good to see there's no apparent regression here.

```diff
diff --git a/src/main.rs b/src/main.rs
index 69a7814..f271880 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -5,5 +5,8 @@ use nexus_rt::write_log;
 
 #[nexus_rt::main]
 fn main() {
-    write_log("Hello, World!\n");
+    let mut i = 0;
+    while i < 10 {
+        i += 1;
+    }
 }
```

Prove then verify with `--impl jolt`:

```
$ cargo nexus prove --impl jolt
   Compiling proc-macro2 v1.0.84
   Compiling unicode-ident v1.0.12
   Compiling syn v1.0.109
   Compiling nexus-rt v1.0.0 (.../nexus-zkvm/only-riscv/runtime)
   Compiling quote v1.0.36
   Compiling nexus-rt-macros v0.1.0 (.../nexus-zkvm/only-riscv/runtime/macros)
   Compiling pr-164 v0.1.0 (/private/tmp/pr-164)
warning: unused import: `nexus_rt::write_log`
...
    Finished release-unoptimized [unoptimized] target(s) in 2.64s
  Preprocessing bytecode ... 533ms
       Finished in 533ms; bytecode size: 55
Executing program...
Executed 104 instructions in 8.011541ms
  Proving program execution ... 2.5s
 Finished in 2.5s
  Saving proof ... 1ms
Finished in 1ms

$ cargo nexus verify --impl jolt
  Preprocessing bytecode ... 480ms
       Finished in 480ms; bytecode size: 55
  Verifying proof ... 143ms
   Finished in 143ms
```